cmake/compiler/gcc/target_riscv.cmake

```diff
@@ -31,4 +32,9 @@ if(CONFIG_COMPRESSED_ISA)
 endif()
 
 list(APPEND TOOLCHAIN_C_FLAGS -mabi=${riscv_mabi} -march=${riscv_march})
-list(APPEND TOOLCHAIN_LD_FLAGS -mabi=${riscv_mabi} -march=${riscv_march})
+
+# The double quotes below ensure that the `-mabi` and `-march` flags are
+# evaluated together by `target_ld_options`. This is necessary because
+# specifying only either one of the flags will cause compilation to fail and
+# that will effectively omit these flags.
+list(APPEND TOOLCHAIN_LD_FLAGS "-mabi=${riscv_mabi} -march=${riscv_march}")
```

Comment:
I see @stephanosio actually did the commit, so he might have an answer.

Reply:
Yes, without the " marks, each option is checked independently, which fails because they must be added together.

Reply:
Yes, except for the fact that the linker option is currently an empty string `""`.
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L1359
```
    zephyr_check_compiler_flag(C "" ${check})
```
so that is why i'm curious exactly what is being solved with this change ?

Reply:
@tejlmand This does solve an error. Reading https://github.com/zephyrproject-rtos/zephyr/commit/5fb5adb34f5672749607a0b652da3e5fff0d6638 would help understand what is going on here.

Reply:
@stephanosio thanks for the link.

And thanks for confirming that this does fix an error.

It might be that the CMake documentation is a bit imprecise on this part (wouldn't be the first time), cause the docs only refers to `CMAKE_REQUIRED_FLAGS` being used for compilation and not linking.

https://cmake.org/cmake/help/latest/module/CheckCCompilerFlag.html
which actually uses `check_c_source_compiles`
https://cmake.org/cmake/help/latest/module/CheckCSourceCompiles.html 
>CMAKE_REQUIRED_FLAGS
>
>    Additional flags to pass to the compiler. ...

However, it could be that as long as we are using gcc based compiler and linker, then it is enough to check the flag on the compiler itself.
And as you confirm this actually fixes an issue you have seen, then i'm fine with taking in this change.

But I have a small concern that the current testing scheme is not complete (another story).
As long as we are using gcc based toolchains for both compiling and linking we might not see any errors.

However, this may change when starting to work with more toolchains, e.g. clang, Metaware, IAR, armclang.
Especially with clang there can be a difference on checking the flags for `ld` or `ldd`, and testing a flag only on the compiler may not be sufficient.

Reply:
@stephanosio we should find another approach for ensuring those flags are checked together.

Quoting flags like this in CMake may have unintended side-effects.
It often works directly on compile options, but if the properties are fetched for other purposes, either using a generator expression or by `get_property(...)` then there is a risk that the flag may be quoted on command line.

One example is `external_lib`:
```
... -fno-common -g "-mabi=ilp32 -march=rv32ima" -imacros ...
riscv64-zephyr-elf-gcc: error: unrecognized argument in option '-mabi=ilp32 -march=rv32ima'
```

Reply:
How about allowing the combined option in TOOLCHAIN_LD_FLAGS and then splitting it into separate options once validated by the compiler:

```diff --git a/cmake/extensions.cmake b/cmake/extensions.cmake
index 94b597051e..98d12e4e4c 100644
--- a/cmake/extensions.cmake
+++ b/cmake/extensions.cmake
@@ -1359,7 +1359,10 @@ function(target_ld_options target scope)
     zephyr_check_compiler_flag(C "" ${check})
     set(CMAKE_REQUIRED_FLAGS ${SAVED_CMAKE_REQUIRED_FLAGS})
 
-    target_link_libraries_ifdef(${check} ${target} ${scope} ${option})
+    string(REPLACE " " ";" options ${option})
+    foreach(suboption IN LISTS options)
+      target_link_libraries_ifdef(${check} ${target} ${scope} ${suboption})
+    endforeach()
   endforeach()
 endfunction()
 
```

Reply:
(Drive-by comment; I haven't been following the whole discussion.)

Sounds like this might be a job for `separate_arguments()`.

```
$ cat ~/foo.cmake
set(foo "-mabi=ilp32 -march=rv32ma --some-option=\"with spaces in the value\"")

separate_arguments(foo_as_list UNIX_COMMAND ${foo})

message(STATUS "foo_as_list=${foo_as_list}")
foreach(arg ${foo_as_list})
  message(STATUS "(${arg})")
endforeach()

$ cmake -P ~/foo.cmake
-- foo_as_list=-mabi=ilp32;-march=rv32ma;--some-option=with spaces in the value
-- (-mabi=ilp32)
-- (-march=rv32ma)
-- (--some-option=with spaces in the value)
```

Reply:
thanks for the reference -- looks like that's a better option than my 'REPLACE' line.

I've pushed an updated picolibc branch with this change.